llm-examples/mpt/mpt-7b/scripts/fine_tune_deepspeed.py

```diff
@@ -0,0 +1,263 @@
+import click
+from functools import partial
+import json
+import logging
+import os
+import numpy as np
+from pathlib import Path
+import torch
+
+from datasets import Dataset, load_dataset
+import transformers
+from transformers import (
+    AutoConfig,
+    AutoModelForCausalLM,
+    AutoTokenizer,
+    DataCollatorForLanguageModeling,
+    PreTrainedTokenizer,
+    Trainer,
+    TrainingArguments,
+    set_seed,
+)
+
+
+os.environ['HF_HOME'] = '/local_disk0/hf'
+os.environ['TRANSFORMERS_CACHE'] = '/local_disk0/hf'
+
+logger = logging.getLogger(__name__)
+
+INTRO_BLURB = "Below is an instruction that describes a task. Write a response that appropriately completes the request."
+INSTRUCTION_KEY = "### Instruction:"
+INPUT_KEY = "Input:"
+RESPONSE_KEY = "### Response:"
+PROMPT_NO_INPUT_FORMAT = """{intro}
+
+{instruction_key}
+{instruction}
+
+{response_key}""".format(
+  intro=INTRO_BLURB,
+  instruction_key=INSTRUCTION_KEY,
+  instruction="{instruction}",
+  response_key=RESPONSE_KEY,
+)
+
+PROMPT_WITH_INPUT_FORMAT = """{intro}
+
+{instruction_key}
+{instruction}
+
+{input_key}
+{input}
+
+{response_key}""".format(
+  intro=INTRO_BLURB,
+  instruction_key=INSTRUCTION_KEY,
+  instruction="{instruction}",
+  input_key=INPUT_KEY,
+  input="{input}",
+  response_key=RESPONSE_KEY,
+)
+
+
+ROOT_PATH = Path(__file__).parent.parent
+MODEL_PATH = "mosaicml/mpt-7b"
+REVISION = "72e5f594ce36f9cabfa2a9fd8f58b491eb467ee7"
+TOKENIZER_PATH = "EleutherAI/gpt-neox-20b"
+DEFAULT_TRAINING_DATASET = "databricks/databricks-dolly-15k"
+CONFIG_PATH = "../../config/a10_config.json"
+LOCAL_OUTPUT_DIR = "/local_disk0/output"
+DEFAULT_SEED = 68
+
+
+def load_training_dataset(
+  tokenizer,
+  path_or_dataset: str = DEFAULT_TRAINING_DATASET,
+  seed: int = DEFAULT_SEED
+  ) -> Dataset:
+    logger.info(f"Loading dataset from {path_or_dataset}")
+    dataset = load_dataset(path_or_dataset)["train"]
+    logger.info(f"Found {dataset.num_rows} rows", )
+
+    def _reformat_data(rec):
+      instruction = rec["instruction"]
+      response = rec["response"]
+      context = rec.get("context")
+
+      if context:
+        questions = PROMPT_WITH_INPUT_FORMAT.format(instruction=instruction, input=context)
+      else:
+        questions = PROMPT_NO_INPUT_FORMAT.format(instruction=instruction)
+      return {"text": f"{{ 'prompt': {questions}, 'response': {response} }}"}
```

Comment:
The `{{ 'prompt': ..., 'response': ... }}` format might not be ideal, because after fine-tuning I find that the model often outputs redundant `}` characters at the end.

Reply:
I see! Maybe it works for MPT models and just doesn't work for the Falcon models. I was using the same script for Falcon so I had the comment.